app/src/main/java/com/duckduckgo/app/browser/WebNavigationState.kt

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright (c) 2019 DuckDuckGo
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.duckduckgo.app.browser
+
+import android.webkit.WebBackForwardList
+import androidx.core.net.toUri
+import com.duckduckgo.app.global.isHttpsVersionOfUri
+
+
+interface WebNavigationState {
+    val originalUrl: String?
+    val currentUrl: String?
+    val title: String?
+    val stepsToPreviousPage: Int
+    val canGoBack: Boolean
+    val canGoForward: Boolean
+    val hasNavigationHistory: Boolean
+}
+
+sealed class WebNavigationStateChange
+data class NewPage(val url: String, val title: String?) : WebNavigationStateChange()
+data class UrlUpdated(val url: String) : WebNavigationStateChange()
+object PageCleared : WebNavigationStateChange()
+object Unchanged : WebNavigationStateChange()
+object Other : WebNavigationStateChange()
+
+fun WebNavigationState.compare(previous: WebNavigationState?): WebNavigationStateChange {
+
+    if (this == previous) {
+        return Unchanged
+    }
+
+    if (originalUrl == null && previous?.originalUrl != null) {
+        return PageCleared
+    }
+
+    val latestUrl = currentUrl ?: return Other
+
+    // A new page load is identified by the original url changing
+    if (originalUrl != previous?.originalUrl) {
+        return NewPage(latestUrl, title)
+    }
+
+    // The most up-to-date record of the url is the current one, this may change during a page load
+    if (currentUrl != previous?.currentUrl) {
+        return UrlUpdated(latestUrl)
+    }
+
+    return Other
+}
+
+data class WebViewNavigationState(private val stack: WebBackForwardList) : WebNavigationState {
+
+    override val originalUrl: String? = stack.originalUrl
+
+    override val currentUrl: String? = stack.currentUrl
+
+    override val title: String? = stack.currentItem?.title
+
+    override val stepsToPreviousPage: Int = if (stack.isHttpsUpgrade) 2 else 1
+
+    override val canGoBack: Boolean = stack.currentIndex >= stepsToPreviousPage
+
+    override val canGoForward: Boolean = stack.currentIndex + 1 < stack.size
+
+    override val hasNavigationHistory = stack.size != 0
+
+    override fun equals(other: Any?): Boolean {
```

Comment:
Why do we need the overrides for `equals` and `hashCode`? can you add a comment explaining the reason if it's needed?

Reply:
Normally the system provided equals is perfect for data objects but the private stack property (which we don't care about) messes that up here. I have added comments